pymc/distributions/distribution.py

```diff
@@ -622,13 +683,28 @@ def dist(
         if logcdf is None:
             logcdf = default_not_implemented(class_name, "logcdf")
 
+        def dist_moment(rv, size, *dist_params):
+            fgraph = rv.owner.op.fgraph.clone()
+            replace_moments = MomentRewrite()
+            replace_moments.rewrite(fgraph)
+            # we need to replace dummy variable by actual dist params
+            for i, par in enumerate([size] + list(dist_params)):
+                fgraph.replace(fgraph.inputs[i], par)
+            # we need to replace dymmy random generators in Scan node inputs
+            for node in fgraph.toposort():
+                if isinstance(node.op, Scan):
```

Comment:
Why do we need a special logic for Scan inside an OpFromGraph?

Reply:
@ricardoV94 After cloning fgraph we have dummy placeholders in `Scan` inner graph inputs for `RandomGenerator`, and we need to replace it by real random generator to avoid error later when we will try to calculate moment. Do we have another way to do it?

Reply:
I see, so the problem is that the scan moment rewrite does not remove the useless updates. That's probably okay. The issue is that we filtered out the RNGs!!! If we had them the replacement would have worked fine. I was able to simplify the function to this and it passes the test that has a scan:

```python
        def dist_moment(rv, *args):
            node = rv.owner
            op = node.op
            rv_out_idx = node.outputs.index(rv)
            
            fgraph = op.fgraph.clone()
            replace_moments = MomentRewrite()
            replace_moments.rewrite(fgraph)
            # Replace dummy inner inputs by outer inputs
            fgraph.replace_all(tuple(zip(op.inner_inputs, node.inputs)), import_missing=True)
            moment = fgraph.outputs[rv_out_idx]
            return moment
```

To tidy up the code, we should dispatch this moment function on `CustomSymbolicDistRV` outside of the `_CustomSymbolicDist` class. Then we only dispatch a user provided moment when that's given like we do now with `if logp is not None:`

You can remove the moment related stuff that now exists in `_CustomSymbolicDist.dist`

Reply:
@ricardoV94 Great, thank you! Will do